python/social_bot/tasks.py

```diff
@@ -968,21 +975,21 @@ def run(self):
             r_contact = self._agent.get_contacts('finger_cnta_r',
                                                  self._object_collision_name)
             # check distance and contacts
-            dist = np.linalg.norm(np.array(obj_pos) - goal_pos)
-            palm_dist = np.linalg.norm(
-                np.array(obj_pos) - np.array(finger_pos))
-            gripping_feature_num = 1.0 * l_contact + 1.0 * r_contact + 10.0 * min(
-                obj_pos[2], 0.1)  # encourge to lift the object
-            gripping = (gripping_feature_num >= 2.5)
+            obj_dist = np.linalg.norm(np.array(obj_pos) - goal_pos)
+            palm_dist = np.linalg.norm(np.array(obj_pos) - np.array(finger_pos))
+            obj_lifted = obj_height / self._obj_init_height - 1.0
+            gripping_feature = 0.25 * l_contact + 0.25 * r_contact + min(
+                obj_lifted, 1.0)  # encourge to lift the object by obj_height
+            gripping = (gripping_feature >= 0.999999)
             # yield rewards
-            if gripping and dist < self._success_distance_thresh:
+            if gripping and obj_dist - self._obj_init_height < self._success_distance_thresh:  # minus an offset of object height
```

Comment:
obj_dist without minus init_height is good enough?  Why do we need this minus?

Reply:
Gripping feature require the agent grasp the object off the ground by at least an init_height (e.g., 2.5cm for 5cm cube) to make sure it is really off the ground. Other wise (like if we are using 0.5*init_height) the agent is more easier to learn to lean the cube to touching the ground with one of its corner, which could also make the center height lifted by some 1.5~2cm and considered as grasped. I've tried some other griping conditions like, does not work. This is the simplest way I can find by lifting it a init_height.

So it is higher than goal indicator by 2.5cm when gripping the object. which makes the _success_distance_thresh smaller by 2.5cm. If we set it to 2.5cm then it will never succeed. To avoid this, there is the minus operation.

Reply:
I think what I don't understand is why (obj_dist - height).  The two are in different dimensions?  obj_dist is on the x-y plane, and height is on z-axis?

Reply:
obj_dist is on 3D xyz dimension now. Perhaps change obj_dist to x-y plane and separate the condition to axis of xy and z will be better. l'll change this.